contracts/fee_currencies.go

```diff
@@ -95,50 +91,52 @@ func CreditFees(
 		return err
 	}
 
-	// Solidity: function transfer(address to, uint256 amount) returns(bool)
-	transfer1Data, err := abi.Pack("transfer", feeHandler, baseTxFee)
-	if err != nil {
-		return fmt.Errorf("pack transfer base fee: %w", err)
-	}
-	_, leftoverGas, err = evm.Call(caller, *feeCurrency, transfer1Data, leftoverGas, big.NewInt(0))
+	leftoverGas, err = transferErc20(evm, abi, caller, feeCurrency, &baseFeeReceiver, baseFee, leftoverGas)
 	if err != nil {
-		return fmt.Errorf("call transfer base fee: %w", err)
+		return fmt.Errorf("base fee: %w", err)
 	}
 
-	if feeRecipient != common.ZeroAddress && tipTxFee.Cmp(common.Big0) == 1 {
-		transfer2Data, err := abi.Pack("transfer", feeRecipient, tipTxFee)
-		if err != nil {
-			return fmt.Errorf("pack transfer tip: %w", err)
-		}
-		_, leftoverGas, err = evm.Call(caller, *feeCurrency, transfer2Data, leftoverGas, big.NewInt(0))
-		if err != nil {
-			return fmt.Errorf("call transfer tip: %w", err)
-		}
-	}
-
-	if refund.Cmp(common.Big0) == 1 {
-		transfer3Data, err := abi.Pack("transfer", from, refund)
-		if err != nil {
-			return fmt.Errorf("pack transfer refund: %w", err)
-		}
-		_, leftoverGas, err = evm.Call(caller, *feeCurrency, transfer3Data, leftoverGas, big.NewInt(0))
+	if tipReceiver != common.ZeroAddress && feeTip.Cmp(common.Big0) == 1 {
+		leftoverGas, err = transferErc20(evm, abi, caller, feeCurrency, &tipReceiver, feeTip, leftoverGas)
 		if err != nil {
-			return fmt.Errorf("call transfer refund: %w", err)
+			return fmt.Errorf("fee tip: %w", err)
 		}
+	} else {
+		// If the coinbase is the zero address, instead refund the tip to the transaction sender
+		refund.Add(refund, feeTip)
 	}
 
 	if l1DataFee != nil {
```

Comment:
There is an inconsistency in how we work with fee values, here we check it for `nil`, in other cases we compare it to `common.Big0` assuming non-nil values. It's unlikely that it will ever be 0 but maybe worth unifying to avoid a possible issues.

Reply:
The difference is that `l1Cost` can be `nil`, because the calculation is optional depending on the chain config. See https://github.com/celo-org/op-geth/blob/1c014101a74464b545a921428c0e2d2d2371b8e0/core/state_transition.go#L683-L689